### PR TITLE
rocr/aie: let a caller declare an argument it synchronizes itself

### PR DESCRIPTION
**Pull request by erwei-xilinx**
## What

Every AIE dispatch walks each argument's declared range with `CLFLUSH` — before the submit and again after (`FlushArguments`, twice per packet in `XdnaDriver::SubmitCmdChain`). That is the right default: it is what makes a host write visible to the agent and an agent write visible to the host.

But it costs about **18 µs per declared MB**, it is proportional to the argument's size rather than to the work, and it is paid on every dispatch — including for memory the host wrote once and the device has owned ever since.

This documents the kernarg size as *the range the runtime keeps coherent for that argument*, and gives zero the meaning "none — the caller synchronizes this range itself".

## Why it matters

A Llama-3.2-1B decode names ~800 MB per token — 738 MB of weights that never change after load, plus a 64 MB KV cache the device owns — and so spent **~14 ms per token** re-flushing them against host writes that never happen:

| | per dispatch |
|---|---|
| XRT (same design, same machine) | 18.5 ms |
| HSA, before | 33.7 ms |
| HSA, after | 18.1 ms |

End to end that is ~22 → ~52 tok/s, i.e. parity with XRT, which pays nothing here because the weights live in a BO uploaded once and each run passes a handle.

The gap does not move with context length, because the declared sizes do not either.

## Why the size field, and not a fence scope

The packet header already carries acquire/release fence scopes, but they cannot express this: the *same* dispatch needs its three small operands flushed and its two large ones left alone. The choice is per-argument.

The size is also a natural place for it, because it already reaches nothing else — the command BO carries only address lo/hi per argument (`2 * num_kernargs` dwords), so the size bounds this synchronization and nothing more.

## Compatibility

A zero-size argument is meaningless today, so no existing caller can depend on the current behaviour for one. Every non-zero size behaves exactly as before. The two AIE clients I could find (this repo's header consumers: `Xilinx/mlir-aie`, `amd/Triton-XDNA`) are unaffected unless they opt in.

## How it was found

1. A 512 MB region declared at full size costs 26.8 ms; the *same region* declared as 2 MB costs 0.62 ms. So the cost tracks the number in the kernargs, not the allocation.
2. Timers inside the caller's dispatch put all of it in the doorbell store — the completion wait returns in 4 µs. `strace` showed no syscall accounting for it (`EXEC_CMD` is 0.38 ms).
3. Sampling the instruction pointer during the doorbell landed every sample in two spots in `SubmitCmdChain`, both of which disassemble to a `clflush` loop indexed by `FlushCpuCache::cacheline_size`.

## Testing

**Unit** — the selection is factored into `ForEachArgumentSyncRange` so it can be exercised with no AIE agent, no driver and no device, against a recording callback. Seven cases: default sync of every argument, a zero-size argument skipped, zero-size skipped *among* non-zero ones (the case the change exists for), the addr/size halves read in the right order, an empty packet, a null kernarg buffer, and a 16-argument packet.

This adds the first unit-test target in `hsa-runtime`, behind `BUILD_UNIT_TESTS` (OFF by default) and skipped when GoogleTest is absent, so it changes nothing about existing builds:

```
cmake -S runtime/hsa-runtime -B build -DBUILD_UNIT_TESTS=ON && cmake --build build && ctest --test-dir build
```

I verified the tests fail (4 of 7) when the skip is reverted, so they are load-bearing rather than decorative.

**Hardware** — on Strix (`aie2p`), with the client in `amd/Triton-XDNA`: 33.7 → 18.1 ms per dispatch, with the decode's logits **bit-identical to XRT** across thousands of dispatches, and identical with the opt-out on versus off. Turning it off and back on inside one process, seconds apart, moves the dispatch 18.0 ↔ 31.9 ms.

One caveat on what I could verify locally: this box has no ROCm device LLVM, so the trap-handler/blit-shader step of a full `hsa-runtime` build does not run here. The changed translation unit compiles clean and the unit tests build and pass standalone; the hardware numbers above come from the client driving the released runtime, where a declared zero costs exactly one cache line (the flush loop is a `do`/`while`) rather than nothing. With this patch it costs nothing, which is strictly less work than what was measured.

### CODE REVIEW

**Comment by erwei-xilinx** [comment]
Closing: the performance problem this addresses is solved on the client side, so this is not needed.
